File: src/layer.zig
const std = @import("std");

weights: []f64,
last_inputs: []const f64,
outputs: []f64,
weight_grads: []f64, // = [1]f64{0} ** (inputSize * outputSize);
input_grads: []f64, //= [1]f64{0} ** (batchSize * inputSize);
batchSize: usize,
inputSize: usize,
outputSize: usize,

const Self = @This();
//var outputs: [batchSize * outputSize]f64 = [1]f64{0} ** (batchSize * outputSize);
pub fn setWeights(self: *Self, weights: []f64) void {
    self.weights = weights;
}

pub fn readParams(self: *Self, params: anytype) !void {
    _ = try params.read(std.mem.sliceAsBytes(self.weights));
}
pub fn writeParams(self: *Self, params: anytype) !void {
    _ = try params.writeAll(std.mem.sliceAsBytes(self.weights));
}
pub fn init(
    alloc: std.mem.Allocator,
    batchSize: usize,
    inputSize: usize,
    outputSize: usize,
) !Self {
    std.debug.assert(inputSize != 0);
    std.debug.assert(outputSize != 0);
    std.debug.assert(batchSize != 0);
    var weights: []f64 = try alloc.alloc(f64, inputSize * outputSize);
    var prng = std.Random.DefaultPrng.init(123);
    var w: usize = 0;
    while (w < inputSize * outputSize) : (w += 1) {
        weights[w] = prng.random().floatNorm(f64) * 0.2;
    }
    return Self{
        .weights = weights,
        .last_inputs = undefined,
        .outputs = try alloc.alloc(f64, outputSize * batchSize),
        .weight_grads = try alloc.alloc(f64, inputSize * outputSize),
        .input_grads = try alloc.alloc(f64, inputSize * batchSize),
        .batchSize = batchSize,
        .outputSize = outputSize,
        .inputSize = inputSize,
    };
}

pub fn deinitBackwards(self: *Self, alloc: std.mem.Allocator) void {
    //alloc.free(self.last_inputs);
    //alloc.free(self.outputs);
    alloc.free(self.weight_grads);
    alloc.free(self.input_grads);
}

pub fn forward(
    self: *Self,
    inputs: []const f64,
) void {
    std.debug.assert(inputs.len == self.inputSize * self.batchSize);
    var b: usize = 0;
    while (b < self.batchSize) : (b += 1) {
        var o: usize = 0;
        while (o < self.outputSize) : (o += 1) {
            var sum: f64 = 0;
            var i: usize = 0;
            while (i < self.inputSize) : (i += 1) {
                sum += inputs[b * self.inputSize + i] * self.weights[self.outputSize * i + o];
            }
            self.outputs[b * self.outputSize + o] = sum;
        }
    }
    self.last_inputs = inputs;
}

pub fn backwards(
    self: *Self,
    grads: []f64,
) void {
    std.debug.assert(self.last_inputs.len == self.inputSize * self.batchSize);

    //self.input_grads = [1]f64{0} ** (inputSize * batchSize);
    //self.weight_grads = [1]f64{0} ** (inputSize * outputSize);

    @memset(self.input_grads, 0);
    @memset(self.weight_grads, 0);

    var b: usize = 0;
    while (b < self.batchSize) : (b += 1) {
        var i: usize = 0;
        while (i < self.inputSize) : (i += 1) {
            var o: usize = 0;
            while (o < self.outputSize) : (o += 1) {
                self.weight_grads[i * self.outputSize + o] +=
                    (grads[b * self.outputSize + o] * self.last_inputs[b * self.inputSize + i]) / @as(f64, @floatFromInt(self.batchSize));
                self.input_grads[b * self.inputSize + i] +=
                    grads[b * self.outputSize + o] * self.weights[i * self.outputSize + o];
            }
        }
    }
}

pub fn applyGradients(self: *Self) void {
    var i: usize = 0;
    while (i < self.inputSize * self.outputSize) : (i += 1) {
        self.weights[i] -= 0.01 * self.weight_grads[i];
    }
}

File: src/main/java/Layer.java
import java.util.Random;

public class Layer {
    private double[] weights;
    private double[] lastInputs;
    public double[] outputs;
    private double[] weightGrads;
    public double[] input_grads;
    private int batchSize;
    private int inputSize;
    private int outputSize;

    public Layer(int batchSize, int inputSize, int outputSize) {
        this.batchSize = batchSize;
        this.inputSize = inputSize;
        this.outputSize = outputSize;
        this.weights = new double[inputSize * outputSize];
        this.outputs = new double[outputSize * batchSize];
        this.weightGrads = new double[inputSize * outputSize];
        this.input_grads = new double[inputSize * batchSize];
        Random prng = new Random(123);
        for (int i = 0; i < inputSize * outputSize; i++) {
            weights[i] = prng.nextGaussian() * 0.2;
        }
    }

    public void setWeights(double[] weights) {
        this.weights = weights;
    }

    public double[] getWeights() {
        return this.weights;
    }

    public void readParams(java.io.DataInputStream params) throws java.io.IOException {
        for (int i = 0; i < weights.length; i++) {
            weights[i] = params.readDouble();
        }
    }

    public void writeParams(java.io.DataOutputStream params) throws java.io.IOException {
        for (double weight : weights) {
            params.writeDouble(weight);
        }
    }

    public void deinitBackwards() {
        // No equivalent in Java, handled by garbage collector
    }

    public void forward(double[] inputs) {
        assert inputs.length == inputSize * batchSize;
        for (int b = 0; b < batchSize; b++) {
            for (int o = 0; o < outputSize; o++) {
                double sum = 0;
                for (int i = 0; i < inputSize; i++) {
                    sum += inputs[b * inputSize + i] * weights[outputSize * i + o];
                }
                outputs[b * outputSize + o] = sum;
            }
        }
        lastInputs = inputs;
    }

    public void backwards(double[] grads) {
        assert lastInputs.length == inputSize * batchSize;
        java.util.Arrays.fill(input_grads, 0);
        java.util.Arrays.fill(weightGrads, 0);
        for (int b = 0; b < batchSize; b++) {
            for (int i = 0; i < inputSize; i++) {
                for (int o = 0; o < outputSize; o++) {
                    weightGrads[i * outputSize + o] += (grads[b * outputSize + o] * lastInputs[b * inputSize + i]) / batchSize;
                    input_grads[b * inputSize + i] += grads[b * outputSize + o] * weights[i * outputSize + o];
                }
            }
        }
    }

    public void applyGradients() {
        for (int i = 0; i < inputSize * outputSize; i++) {
            weights[i] -= 0.01 * weightGrads[i];
        }
    }

    public int getSize() {
        return this.outputSize;
    }

    public void reinit(double percent) {
        Random prng = new Random(123);
        for (int i = 0; i < inputSize * outputSize; i++) {
            weights[i] = prng.nextGaussian() * 0.2 * percent;
        }
    }
}
